Feat: Bulk Digestion in Gastric Acid

kd8lvt opened the pull request:
Throw digestible item in acid, get nutrient paste out.
Has a 80% efficiency when compared the the digester for balance, but the fact that you have to grow your cradle enough to get acid is also a component.

Fairly simple to change the efficiency on it, it's in AcidInteractions.

Annoyingly, neither Forge nor Minecraft give a clean way to get JUST item entities during a tick, so I had to Mixin into ItemEntity.tick()

kd8lvt committed:
Feat: Bulk Digestion in Gastric Acid

kd8lvt committed:
Feat: Bulk Digestion in Gastric Acid

Elenterius commented on src/main/java/com/github/elenterius/biomancy/mixin/ItemEntityMixin.java:
```diff
@@ -0,0 +1,20 @@
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(ItemEntity.class)
+public class ItemEntityMixin {
```
mixin class should be abstract

Elenterius commented on src/main/java/com/github/elenterius/biomancy/mixin/ItemEntityMixin.java:
```diff
@@ -0,0 +1,20 @@
+public class ItemEntityMixin {
+	@Unique
+	@Inject(at=@At("TAIL"),method={"tick()V"})
+	public void biomancy$tick(CallbackInfo ci) {
```
- remove the unique annotation
- rename the method to onTick
- make the method private

Elenterius commented on src/main/java/com/github/elenterius/biomancy/mixin/ItemEntityMixin.java:
```diff
@@ -0,0 +1,20 @@
+	public void biomancy$tick(CallbackInfo ci) {
+		ItemEntity self = (ItemEntity)((Object)this); //I hate casting like this on so many levels
+		if (self.level().isClientSide()) return;
+		AcidInteractions.tryDigest(self);
```
since we are injecting at the tail we also need to check if the item hasn't been yet removed

 `if (self.isRemoved()) return;`

Elenterius commented on src/main/java/com/github/elenterius/biomancy/init/AcidInteractions.java:
```diff
@@ -157,4 +165,55 @@ else if (livingEntity.tickCount % 10 == 0 && livingEntity.getRandom().nextFloat(
+	@SuppressWarnings("RedundantIfStatement")
+	private static boolean digestible(ItemEntity itemEntity) {
+		if (!itemEntity.isInFluidType(ModFluids.ACID_TYPE.get()) && !itemEntity.level().getBlockState(itemEntity.blockPosition()).is(ModBlocks.ACID_CAULDRON.get())) return false;
+		if (itemEntity.getItem().is(ModItems.NUTRIENT_PASTE.get())) return false;
```
you also need to check for Nutrient Bars

Elenterius replied:
and living flesh needs to be excluded as well

Elenterius commented on src/main/java/com/github/elenterius/biomancy/init/AcidInteractions.java:
```diff
@@ -157,4 +165,55 @@ else if (livingEntity.tickCount % 10 == 0 && livingEntity.getRandom().nextFloat(
+		if (data.contains(TIMER_KEY) && data.getInt(TIMER_KEY) < 100) {
+			data.putInt(TIMER_KEY,data.getInt(TIMER_KEY)+1);
+			//TODO: Bubble particles? Couldn't get it working for some reason.
+			// Copied the addParticle call from handleEntityInsideAcid and it said no.
```
`addParticle` is only implemented by the client level and you are calling it on the server level side

kd8lvt replied:
It also doesn't help that the bubble particle disappears next tick when not in a water block 🤦

Elenterius commented on src/main/java/com/github/elenterius/biomancy/init/AcidInteractions.java:
```diff
@@ -157,4 +165,55 @@ else if (livingEntity.tickCount % 10 == 0 && livingEntity.getRandom().nextFloat(
 
+	private static final String TIMER_KEY = "biomancy:digestion_timer";
+	private static final float EFFICIENCY = 0.8f;
+	public static void tryDigest(ItemEntity itemEntity) {
```
Don't call this method each tick due to the recipe lookup and the possibility of having many item entities in the world inside acid. I would prefer every 10-20 ticks.

Furthermore, cache the found recipe id and use that one first to test if the recipe is still valid.

Elenterius commented on src/main/java/com/github/elenterius/biomancy/init/AcidInteractions.java:
```diff
@@ -157,4 +165,55 @@ else if (livingEntity.tickCount % 10 == 0 && livingEntity.getRandom().nextFloat(
+	}
+
+	private static void digestIntoNutrientPasteStacks(ItemEntity itemEntity) {
+		if (getDigestionRecipe(itemEntity).isEmpty()) return;
```
avoid the unnecessary lookup of the recipe and instead pass it into the method. As it stand you lookup the recipe thrice.

Elenterius commented on src/main/java/com/github/elenterius/biomancy/init/AcidInteractions.java:
```diff
@@ -157,4 +165,55 @@ else if (livingEntity.tickCount % 10 == 0 && livingEntity.getRandom().nextFloat(
+	private static void digestIntoNutrientPasteStacks(ItemEntity itemEntity) {
+		if (getDigestionRecipe(itemEntity).isEmpty()) return;
+		DigestingRecipe recipe = getDigestionRecipe(itemEntity).get();
+		BehavioralInventory<?> inv = BehavioralInventory.createServerContents(1,player->false,()->{});
```
btw you don't need to use this, you can use the RecipeWrapper class instead

kd8lvt replied:
I'm not sure what you mean by this, I still need to make an inventory to instantiate a RecipeWrapper?

Elenterius commented on src/main/java/com/github/elenterius/biomancy/init/AcidInteractions.java:
```diff
@@ -157,4 +165,55 @@ else if (livingEntity.tickCount % 10 == 0 && livingEntity.getRandom().nextFloat(
+		BehavioralInventory<?> inv = BehavioralInventory.createServerContents(1,player->false,()->{});
+		inv.insertItemStack(itemEntity.getItem());
+		ItemStack resultStack = recipe.assemble(inv,itemEntity.level().registryAccess());
+		int amt = (int)Math.floor(resultStack.getCount()*itemEntity.getItem().getCount()*EFFICIENCY);
```
i prefer full variable names

kd8lvt committed:
Feat: Bulk Digestion in Gastric Acid

(Implement requested changes)

kd8lvt commented:
Implemented the changes I resolved. As for the others-
* Particles: I wanted to do them for the visual feedback of *something* happening, but even after creating an entirely different particle that's just a green bubble that doesn't care where it is, (and switching to clientside of course,) still no bueno.
* RecipeWrapper: Still unsure what you mean by this; I'm more than willing to learn, just haven't quite figured out how you'd want me to use/implement that specifically.

kd8lvt committed:
Merge remote-tracking branch 'origin/mc1.20.1/feat/acid-digests-things' into mc1.20.1/feat/acid-digests-things

Elenterius commented:
> RecipeWrapper: Still unsure what you mean by this; I'm more than willing to learn, just haven't quite figured out how you'd want me to use/implement that specifically.

I'm planning to phase out the use of vanilla inventory containers and my extensions of it (BehavioralInventory) in favor of simply using (neo-)forges ItemStackHandler interfaces. The recipe wrapper is just used when you need a vanilla inventory container for assembling recipes, etc.

Example: 

```java
RecipeWrapper inv = new RecipeWrapper(new ItemStackHandler(1));
inv.setItem(0, inputStack);
ItemStack resultStack = recipe.assemble(inv, level.registryAccess());
```

Elenterius commented on src/main/java/com/github/elenterius/biomancy/init/AcidInteractions.java:
```diff
@@ -157,4 +166,59 @@ else if (livingEntity.tickCount % 10 == 0 && livingEntity.getRandom().nextFloat(
+		if (recipe == null) return false;
+		if (!itemEntity.isInFluidType(ModFluids.ACID_TYPE.get()) && !itemEntity.level().getBlockState(itemEntity.blockPosition()).is(ModBlocks.ACID_CAULDRON.get())) return false;
+		//Inside method to prevent missing registry object errors during init
+		Item[] blacklistedItems = {ModItems.NUTRIENT_PASTE.get(),ModItems.NUTRIENT_BAR.get(),ModItems.LIVING_FLESH.get()};
```
Replace this with a static Set **OR** replace this with a new item tag.


Set Example, that is populated inside the `register` method.

```java
public static final Set<Item> NOT_DIGESTIBLE = new HashSet<>();
...

register() {
  ...
  NOT_DIGESTIBLE.put(ModItems.NUTRIENT_PASTE.get);
  ...
}

private static boolean digestible(...) {
    if (NOT_DIGESTIBLE.contains(itemEntity.getItem().getItem()) return false;
}
```

kd8lvt replied:
Ah, an item tag is a good idea.

Elenterius commented on src/main/java/com/github/elenterius/biomancy/init/AcidInteractions.java:
```diff
@@ -157,4 +166,59 @@ else if (livingEntity.tickCount % 10 == 0 && livingEntity.getRandom().nextFloat(
+		return true;
+	}
+
+	private static DigestingRecipe cachedRecipe = null;
```
i should have been more clear what i mean with caching. I meant storing the recipe Id in the persistent data.

Example:
```java
	public static void tryDigest(ItemEntity itemEntity) {
		ItemStack stack = itemEntity.getItem();
		CompoundTag tag = getOrCreateTag(itemEntity.getPersistentData(), DIGESTION_KEY);

		ResourceLocation lastRecipeId = ResourceLocation.tryParse(tag.getString(LAST_RECIPE_KEY));
		Optional<Pair<ResourceLocation, DigestingRecipe>> optionalPair = DigesterBlockEntity.RECIPE_TYPE.get().getRecipeForIngredient(itemEntity.level(), stack, lastRecipeId);

		if (optionalPair.isEmpty()) return;

		ResourceLocation recipeId = optionalPair.get().getFirst();
		DigestingRecipe recipe = optionalPair.get().getSecond();

		tag.putString(LAST_RECIPE_KEY, recipeId.toString());

		int timer = tag.getInt(TIMER_KEY);
		if (timer < 10) {
			tag.putInt(TIMER_KEY, timer + 1);
		}
		else {
			if (!itemEntity.level().isClientSide) {
				RecipeWrapper inventory = new RecipeWrapper(new ItemStackHandler(1));
				inventory.setItem(0, stack);
				digestIntoNutrientPasteStacks(itemEntity, inventory, recipe);
			}
			tag.remove(TIMER_KEY);
		}
	}
```

kd8lvt replied:
Ah ok, I'll do that.
Would that not still need to query the recipe every time it is run, though? Saying as you're not storing a reference to the recipe itself?

Elenterius reviewed (changes requested):
don't forget to format your file ;)

Elenterius commented:
>  Particles: I wanted to do them for the visual feedback of _something_ happening, but even after creating an entirely different particle that's just a green bubble that doesn't care where it is, (and switching to clientside of course,) still no bueno.

no idea whats going on with that but if you feel like it you can include the removed code for the particle in the PR (or link to a branch with the code) and i will look into it at a later point. I'm mentioning this because i do want to spruce up acid fluid with bubbles.

kd8lvt commented:
> > Particles: I wanted to do them for the visual feedback of _something_ happening, but even after creating an entirely different particle that's just a green bubble that doesn't care where it is, (and switching to clientside of course,) still no bueno.
> 
> no idea whats going on with that but if you feel like it you can include the removed code for the particle in the PR (or link to a branch with the code) and i will look into it at a later point. I'm mentioning this because i do want to spruce up acid fluid with bubbles.

Will do, I've been busy with IRL stuff the last week, hence why I haven't touched it lol

Edit: I got them working in the latest batch of commits. Probably related to #151 - particles don't render through Acid without Fabulous graphics, which pretty thoroughly ruins the effect :(

kd8lvt committed:
Merge remote-tracking branch 'refs/remotes/upstream/mc1.20.1/dev' into mc1.20.1/feat/acid-digests-things

# Conflicts:
#	src/main/resources/mixins.biomancy.json

kd8lvt committed:
Feat: Bulk Digestion in Gastric Acid

Implement requested changes
Note: Particles (even Vanilla ones) don't render under/through Acid without Fabulous graphics enabled. Likely related to [#151](https://github.com/Elenterius/Biomancy/issues/151)

kd8lvt committed:
Feat: Bulk Digestion in Gastric Acid

Implement requested changes
Note: Particles (even Vanilla ones) don't render under/through Acid without Fabulous graphics enabled. Likely related to [#151](https://github.com/Elenterius/Biomancy/issues/151)

kd8lvt committed:
Feat: Bulk Digestion in Gastric Acid

Implement requested changes
Note: Particles (even Vanilla ones) don't render under/through Acid without Fabulous graphics enabled. Likely related to [#151](https://github.com/Elenterius/Biomancy/issues/151)

kd8lvt committed:
Merge remote-tracking branch 'origin/mc1.20.1/feat/acid-digests-things' into mc1.20.1/feat/acid-digests-things

kd8lvt commented:
We love redundant merging! ([6f01210](https://github.com/Elenterius/Biomancy/pull/150/commits/6f012100384f00d61af219b6c5eb933801234f11))
Woops.

Elenterius committed:
Merge branch 'mc1.20.1/dev' into mc1.20.1/feat/acid-digests-things
